### PR TITLE
Faster wkw integration

### PR DESCRIPTION
**Pull request by normanrz**
This PR adds a rust module that implements the loading of WKW buckets in an optimized way. The rust module supports only reading of individual buckets and strips all the code for arbitrary cutouts away. Memory-mapping is used for IO. There is an lru cache for the memory-mapped file handles, which helps to minimize syscalls. The Python interface is built with asyncio. All data loading is off-loaded to a thread pool for parallelization and to avoid blocking Python's main thread.

Depends on #110 and https://github.com/scalableminds/webknossos-wrap/pull/66

### CODE REVIEW

File: fast_wkw/src/lib.rs
Comment (valentin-pinkau) [inline fast_wkw/src/lib.rs:71]: Why do we need tokio::task::spawn_blocking?
Couldn't we just do `dataset.read_block(...).await` (as suggested below)?

File: fast_wkw/src/wkw_dataset.rs
Comment (valentin-pinkau) [inline fast_wkw/src/wkw_dataset.rs:82]: What do you think about making this function async with `file.read().unwrap().read_block(file_src_pos, buf).await`

**Comment by valentin-pinkau** [comment]
Very cool stuff! I think it makes totally sense to have a low latency wkw interfaces which operates on buckets!

File: fast_wkw/src/wkw_dataset.rs
Comment (normanrz) [reply fast_wkw/src/wkw_dataset.rs:82]: I think that conflicts with the memory-mapped file approach. Reading from disk by a page fault is a blocking operation afaik. Putting everything in a thread pool should help make that non-blocking. Lz4 decoding is technically also a blocking operation, but not as problematic because it is very quick.

File: wkconnect/backends/wkw/models.py
Comment (fm3) [inline wkconnect/backends/wkw/models.py:29]: I assume this repository is a cache? Is there a specific reason for this name? I confused it with https://github.com/scalableminds/webknossos-connect/blob/master/wkconnect/repository.py – maybe this could just be called something with wkw cache?

File: wkconnect/routes/datasets/histogram.py
Comment (fm3) [inline wkconnect/routes/datasets/histogram.py:37]: Maybe this could be extracted into a function?
Also, could it be skipped for backends that are not wkw?

File: wkconnect/routes/datasets/histogram.py
Comment (fm3) [inline wkconnect/routes/datasets/histogram.py:49]: 🎉

File: wkconnect/backends/wkw/backend.py
Comment (fm3) [inline wkconnect/backends/wkw/backend.py:22]: What’s the unit here? File handles? Bytes? Buckets? Could this be encoded in the config key?

File: Dockerfile
Comment (fm3) [inline Dockerfile:29]: Does any of this have to happen for a local non-docker installation too? If so, could you adapt the readme? It currently only says poetry install

File: .github/workflows/main.yml
Comment (fm3) [inline .github/workflows/main.yml:28]: Should probably be a pinned release of wkw once it exists?

File: Dockerfile
Comment (fm3) [reply Dockerfile:29]: Also I noticed this is duplicated in Dockerfile and the CI yaml. Maybe it could be extracted into a script?

**Review by fm3 — CHANGES_REQUESTED** [review]
Great stuff! I’m looking forward to seeing this in action. Did you measure the perf impact?
I mostly looked through the python code since I am not familiar enough with rust to say much about that part at this time. See my minor comments on python and infrastructure :)

File: Dockerfile
Comment (normanrz) [reply Dockerfile:29]: This will be removed once scalableminds/webknossos-wrap#66 is merged. Then, the regular rust package manager will load wkwrap without any special code.

File: wkconnect/routes/datasets/histogram.py
Comment (normanrz) [reply wkconnect/routes/datasets/histogram.py]: > Also, could it be skipped for backends that are not wkw?

We could. But the mag-detection part would still be required. And perf may be better with 32-aligned requests (because most backends have chunked storage with a 2**x ≥ 32 size). I don't see any harm in aligned the requests.

**Review by fm3 — APPROVED** [review]
Python part LGTM :) Let’s wait for the wk-wrap PR and Valentin’s final approval

**Review by valentin-pinkau — CHANGES_REQUESTED** [review]
Very cool! It would be very interesting to see how the performance characteristics change with this new implementation e.g. (Latency of a cached bucket, avg. response time for random access, ...).
In addition, I am not sure about the benefit of the async DatasetHandle.read_block interface. My fear is that it imposes more overhead than wins due to otherwise blocked threads. Do you think you could give some numbers on that?

File: fast_wkw/src/wkw_dataset.rs
Comment (normanrz) [reply fast_wkw/src/wkw_dataset.rs:10]: Good point

**Comment by normanrz** [comment]
> It would be very interesting to see how the performance characteristics change with this new implementation e.g. (Latency of a cached bucket, avg. response time for random access, ...).

I think the interesting comparison will be between this implementation and the Scala implementation. The previous implementation with the regular Python wkw lib was prohibitively slow.

> In addition, I am not sure about the benefit of the async DatasetHandle.read_block interface. My fear is that it imposes more overhead than wins due to otherwise blocked threads. Do you think you could give some numbers on that?

I think there is no way around async, here. The python http server is an async server that runs on a single threaded event loop. We need a thread pool of some sorts to not block the whole server when reading buckets from disk. Remember that most datasets are stored on spinning disks. Using a thread pool, as implemented here, is a typical pattern in other event loop systems such as node.js. Afaik the communication between the threads in tokio are implemented with message queues, so, if at all, it would hurt single-bucket latency and not necessarily throughput.

I don't have numbers for either :-( I think this implementation is fast enough :tm: to be merged, because it is much faster than the previous wkw impl. We should do end-to-end tests on one of our production deployments to test wkconnect vs Scala datastore at some point.

**Review by valentin-pinkau — APPROVED** [review]
Alright, makes sense to me. So I am curious how this compares to the scala implementation :)

**Pull request merged**
